File: DivideTwoIntegers.py
class Solution:
    # @return an integer
    def divide(self, dividend, divisor):
        flag = +1
        if dividend > 0 and divisor < 0 or dividend < 0 and divisor > 0:
            if abs(dividend) < abs(divisor):
                return 0
                flag = -1
                
        res = 0
        a, b = abs(dividend), abs(divisor)
        while a >= b:
            sum, cnt = b, 1
            while sum + sum < a:
                sum += sum
                cnt += cnt
            a -= sum
            res += cnt
        if flag == -1:
            return -res
        return res

    # binary operator
    def divide2(self, dividend, divisor):
        flag = +1
        if dividend > 0 and divisor < 0 or dividend < 0 and divisor > 0:
            if abs(dividend) < abs(divisor):
                return 0
        flag = -1
                    
        res = 0
        a, b = abs(dividend), abs(divisor)
        while a >= b:
            sum, cnt = b, 1
            while sum<<1 < a:
                sum <<= 1
                cnt <<= 1
            a -= sum
            res += cnt
        if flag == -1:
            return -res
        return res

if __name__ == '__main__':
    so = Solution()
    print so.divide2(43421,3244)

File: CPP/Second/divide_two_integers.cc
// https://oj.leetcode.com/problems/divide-two-integers/

// Divide two integers without using multiplication, division and mod operator.

#include <iostream>
#include <algorithm>
#include <vector>
using namespace std;

class Solution {
public:
    int divide(int dividend, int divisor) {
        bool neg = false;
	 long dvd = dividend;
	 long dvr = divisor;

        if (dvd < 0) {
            dvd = -dvd; 
            neg = !neg;
        }
        if (dvr < 0) {
            dvr = -dvr;
            neg = !neg;
        }

        long long res = (dvr == 0) ? INT_MAX : 0;
        int times = 0;
        
        while ((dvd>>1) >= dvr) {
            dvr <<= 1;
            times++;
        }
        while (times >= 0) {
            if (dvd >= dvr) {
                res += (1 << times);
                dvd -= dvr;
            }
            dvr >>= 1;
            times--;
        }
	res = neg ? -res : res;
         if (res > INT_MAX || res < -INT_MAX)
             return INT_MAX;
        return res;
    }
};
int main(int argc, char *argv[])
{
    Solution so;
    cout<<so.divide(-1010369383, -2147483648)<<endl;
    cout<<so.divide(-2147483648, 1)<<endl;
    cout<<so.divide(-2147483648, -1)<<endl;
    cout<<"------------"<<endl;
    cout<<INT_MAX<<endl;
    cout<<INT_MIN<<endl;
    cout<<-INT_MAX<<endl;
    return 0;
}
